datafusion/physical-plan/src/aggregates/group_values/group_column.rs

```diff
@@ -287,6 +469,63 @@ where
         };
     }
 
+    fn vectorized_equal_to(
```

Comment:
In hash join implementation the equal to is basically using the arrow kernels `take` (for rows to check) + `eq` + `and` (if multiple columns), which is I think pretty fast, at least for primitives (and string views, byte/string might be slower as it needs some more copying). This performed better than the row-by-row check.
I didn't look to much at this implementation, but does it seem worth trying to you?

Reply:
🤔 Yes, I agree with the row by row checking is indeed not efficient enough, and switching the similar implementation in `hash_join` may be really worth trying.

Maybe better to try it in the follow on pr? Following points are still not clear for me, and I want to experiment about them:
- If we need a reusable buffer to hold the taken values?
- It seems that the `cmp` for some arrays like `StringArray` and `StringViewArray`  is expansive? 
   Is it better to just check row by row for skipping some unnecessary checkings (if `row` not equal in `col a`, actually we don't need to check it again in `col b`)?

Reply:
At least `take` for `StringArray` is expensive as it needs to copy the string data, recheck the offsets, etc. For string / byte views this will probably be much better.
I think some approach for using `take` + `eq` for primitives while using a custom one for string arrays will likely be best overal.

Even better would be to write a optimized implementation (an arrow kernel that performs equality of two arrays by indices efficiently while applying the same tricks done by take / eq kernels). The join implementation can benefit from this as well.

Reply:
> At least `take` for `StringArray` is expensive as it needs to copy the string data, recheck the offsets, etc. For string / byte views this will probably be much better. I think some approach for using `take` + `eq` for primitives while using a custom one for string arrays will likely be best overal.
> 
> Even better would be to write a optimized implementation (an arrow kernel that performs equality of two arrays by indices efficiently while applying the same tricks done by take / eq kernels). The join implementation can benefit from this as well.

Make sense to me, and support it in `arrow` seems really good.

I plan to sort out codes in this pr , and then try to improve the primitive case using `take` + `eq`.
I guess some case like `q32` in clickbench is possible to get benefit from it.

Reply:
> At least `take` for `StringArray` is expensive as it needs to copy the string data, recheck the offsets, etc. For string / byte views this will probably be much better. I think some approach for using `take` + `eq` for primitives while using a custom one for string arrays will likely be best overal.
> 
> Even better would be to write a optimized implementation (an arrow kernel that performs equality of two arrays by indices efficiently while applying the same tricks done by take / eq kernels). The join implementation can benefit from this as well.

Make sense to me, and support it in `arrow` seems really good.

I plan to sort out codes in this pr , and then try to improve the primitive case using `take` + `eq`.
I guess some case like `q32` in clickbench is possible to get benefit from it.

Reply:
Hmm... 🤔  I think `filter` is a subset of `take` in what it supports as you can repeat values / use them out of order in `take` but can't do that in `filter`.

Somehow generalizing them sounds reasonable, as `filter` even uses the same strategy whenever the predicate is sparse (converting to indices).

Appending to a in-progress array seems like it could be very useful to aggregates yeah, I think that will mostly remove the need for the `GroupColumn` code...

For the equality check however we would need to take in two arrays of indices and/or boolean predicate followed by a equality operation (ideally in one go without copying the input arrays).

Reply:
What i have been dreaming about with @XiangpengHao  is maybe something like adding `take` / `filter` to arrow array builders

I took this opportunity to write up the idea (finally) for your amusement: 
- https://github.com/apache/arrow-rs/issues/6692

Reply:
> At least `take` for `StringArray` is expensive as it needs to copy the string data, recheck the offsets, etc. For string / byte views this will probably be much better. I think some approach for using `take` + `eq` for primitives while using a custom one for string arrays will likely be best overal.
> 
> Even better would be to write a optimized implementation (an arrow kernel that performs equality of two arrays by indices efficiently while applying the same tricks done by take / eq kernels). The join implementation can benefit from this as well.

I try `take + eq` approach , but see no improvement... code is here https://github.com/Rachelint/arrow-datafusion/tree/optimize-vectorized-operations-bak